Add project details view toggle to overview

spadgett opened the pull request:
![project-dashboard](https://user-images.githubusercontent.com/1167259/48293260-3c975c00-e44c-11e8-9d72-8b6ad071ac54.png)

/assign @TheRealJon 
@openshift/team-ux-review

TheRealJon commented:
I'm not sure I like the idea of switching the view at the top level like this. As you showed in your demo, it's possible to have the details side panel open for a resource on the overview list, then switch to the project details view, which totally changes the context, and the details panel no longer makes any sense. IMO, this might need some more design consideration. 

One solution I was considering is that we could use the horizontal nav that we use on almost every other page to show 'Project Details' and 'Topology' tabs. This could keep the project details and topology info separated into distinct pages.

Another idea was to change the 'Overview' nav item to 'Dashboard' and make that the new landing page. This would show the project details and all of the metrics and visualizations we were planning on adding. We could then add a 'Topology' nav item which would be what is currently called the 'Overview'. I brought this up before, but whether we are showing the topology info in list or graph form, it is still topology information, and that term could be used to describe either view. Our view switcher would then just be 'Topology List' or 'Topology Graph' or something along those lines.

@spadgett @serenamarie125 thoughts?

spadgett commented:
@serenamarie125 I made the changes we talked about. I think it's much better than before.

* Moved view selector just to the right of the header. This fixes some of the weirdness with the toggles moving and the sidebar.
* Renamed nav item "Status"

@TheRealJon

spadgett commented:
<img width="1147" alt="screen shot 2018-11-06 at 4 03 39 pm" src="https://user-images.githubusercontent.com/1167259/48093615-d6a59d00-e1dd-11e8-8c7a-a7d6ea199148.png">

spadgett commented:
Rebased

spadgett commented:
Updated. @TheRealJon PTAL

spadgett commented:
Rebased to the latest overview changes.

spadgett commented:
With the latest changes:

![project-overview-dashboard](https://user-images.githubusercontent.com/1167259/48293223-08bc3680-e44c-11e8-8340-501fe68b1cd1.png)

spadgett commented:
jenkins rebuild

beanh66 commented:
@spadgett This looks great with the latest dashboard changes to include "card" borders around each section 👍

TheRealJon commented on frontend/public/components/namespace.jsx:
```diff
@@ -5,12 +5,13 @@ import { Tooltip } from './utils/tooltip';
 import * as fuzzy from 'fuzzysearch';
 
-import { NamespaceModel, ProjectModel, SecretModel } from '../models';
+import { NamespaceModel, ProjectModel, RoleBindingModel, SecretModel } from '../models';
```
nit: break long import lists into multiple lines and sort.

TheRealJon replied:
Also, I don't know if we want to make this a convention, but I've been sorting multi-line imports last in the overall list of imports. I've also been breaking the imports into three separately sorted sections: external imports, ancestor imports, and sibling imports. Since some components are now nested further into the file structure, the three sections make it easier to quickly see where dependencies are coming from.

TheRealJon commented on frontend/public/components/namespace.jsx:
```diff
@@ -5,12 +5,13 @@ import { Tooltip } from './utils/tooltip';
 import { k8sGet } from '../module/k8s';
 import { formatNamespacedRouteForResource, UIActions } from '../ui/ui-actions';
+import { Health } from './graphs/health';
 import { ColHead, DetailsPage, List, ListHeader, ListPage, ResourceRow } from './factory';
```
nit/not new: break long import lists into multiple lines and sort.

TheRealJon commented on frontend/public/components/namespace.jsx:
```diff
@@ -5,12 +5,13 @@ import { Tooltip } from './utils/tooltip';
 import { ColHead, DetailsPage, List, ListHeader, ListPage, ResourceRow } from './factory';
 import { SafetyFirst } from './safety-first';
-import { ActionsMenu, Cog, Dropdown, Firehose, LabelList, LoadingInline, navFactory, ResourceCog, SectionHeading, ResourceLink, ResourceSummary, humanizeMem, MsgBox } from './utils';
+import { ActionsMenu, Cog, Dropdown, Firehose, LabelList, LoadingInline, navFactory, ResourceCog, SectionHeading, ResourceIcon, ResourceLink, ResourceSummary, humanizeMem, MsgBox, resourceListPathFromModel } from './utils';
```
nit: break long import lists into multiple lines and sort.

TheRealJon commented on frontend/public/components/namespace.jsx:
```diff
@@ -5,12 +5,13 @@ import { Tooltip } from './utils/tooltip';
 import { SafetyFirst } from './safety-first';
-import { ActionsMenu, Cog, Dropdown, Firehose, LabelList, LoadingInline, navFactory, ResourceCog, SectionHeading, ResourceLink, ResourceSummary, humanizeMem, MsgBox } from './utils';
+import { ActionsMenu, Cog, Dropdown, Firehose, LabelList, LoadingInline, navFactory, ResourceCog, SectionHeading, ResourceIcon, ResourceLink, ResourceSummary, humanizeMem, MsgBox, resourceListPathFromModel } from './utils';
 import { createNamespaceModal, createProjectModal, deleteNamespaceModal, configureNamespacePullSecretModal } from './modals';
```
nit/not new: break long import lists into multiple lines and sort.

TheRealJon commented on frontend/public/components/namespace.jsx:
```diff
@@ -158,62 +168,110 @@ class PullSecret extends SafetyFirst {
   </div>;
 };
 
+const OverviewHealth = requirePrometheus(({ns}) => <div className="group">
```
Move overview related components into `frontend/public/components/overview/namespace-overview.tsx`

TheRealJon commented on frontend/public/components/overview/index.tsx:
```diff
@@ -243,27 +279,41 @@ const OverviewHeading: React.SFC<OverviewHeadingProps> = ({disabled, groupOption
-class OverviewDetails extends SafetyFirst<OverviewDetailsProps, OverviewDetailsState> {
+const OverviewHeading = connect<OverviewHeadingPropsFromState, OverviewHeadingPropsFromDispatch, OverviewHeadingOwnProps>(headingStateToProps, headingDispatchFromProps)(OverviewHeading_);
+
+const mainContentStateFromProps = ({UI}): OverviewMainContentPropsFromState => {
```
```suggestion
const mainContentStateToProps = ({UI}): OverviewMainContentPropsFromState => {
```

TheRealJon commented on frontend/public/components/overview/index.tsx:
```diff
@@ -718,28 +773,33 @@ class OverviewDetails extends SafetyFirst<OverviewDetailsProps, OverviewDetailsS
   }
 }
 
+const OverviewMainContent = connect<OverviewMainContentPropsFromState, OverviewMainContentPropsFromDispatch, OverviewMainContentOwnProps>(mainContentStateFromProps, mainContentDispatchToProps)(OverviewMainContent_);
```
```suggestion
const OverviewMainContent = connect<OverviewMainContentPropsFromState, OverviewMainContentPropsFromDispatch, OverviewMainContentOwnProps>(mainContentStateToProps, mainContentDispatchToProps)(OverviewMainContent_);
```

TheRealJon commented on frontend/public/components/overview/index.tsx:
```diff
@@ -219,19 +223,51 @@ const sortBuilds = (builds: K8sResourceKind[]): K8sResourceKind[] => {
+  return { selectedView };
+};
+
+const headingDispatchFromProps = (dispatch): OverviewHeadingPropsFromDispatch => ({
```
```suggestion
const headingDispatchToProps = (dispatch): OverviewHeadingPropsFromDispatch => ({
```

TheRealJon commented on frontend/public/components/overview/index.tsx:
```diff
@@ -243,27 +279,41 @@ const OverviewHeading: React.SFC<OverviewHeadingProps> = ({disabled, groupOption
 );
 
-class OverviewDetails extends SafetyFirst<OverviewDetailsProps, OverviewDetailsState> {
+const OverviewHeading = connect<OverviewHeadingPropsFromState, OverviewHeadingPropsFromDispatch, OverviewHeadingOwnProps>(headingStateToProps, headingDispatchFromProps)(OverviewHeading_);
```
```suggestion
const OverviewHeading = connect<OverviewHeadingPropsFromState, OverviewHeadingPropsFromDispatch, OverviewHeadingOwnProps>(headingStateToProps, headingDispatchToProps)(OverviewHeading_);
```

spadgett commented:
@TheRealJon Thanks, updated

We might tackle import styles separately, although the ones you commented on went away when I moved code around.

TheRealJon commented on frontend/public/components/overview/index.tsx:
```diff
@@ -243,27 +279,41 @@ const OverviewHeading: React.SFC<OverviewHeadingProps> = ({disabled, groupOption
   metricsInterval_: any;
 
-  constructor(props: OverviewDetailsProps) {
+  constructor(props: OverviewMainContentProps) {
```
No change necessary for this PR, but I think I got a little over-zealous with the type definitions here. These would be covered by the class definition above.

TheRealJon commented:
/lgtm
